### PR TITLE
Fixed #31606 -- Allowed using condition with lookups in When() expression.

### PR DESCRIPTION
**Pull request by rheard**
Modify When to allow for a condition argument mixed with condition keywords.

Add test for using Exists in When, also add test for using condition argument with condition keyword.

### CODE REVIEW

**Comment by charettes** [comment]
This is a patch for https://code.djangoproject.com/ticket/31606

File: tests/expressions_case/tests.py [hunk 1]
```diff
@@ -312,6 +312,30 @@ def test_in_subquery(self):
...
+    def test_exists(self):
+        instance = CaseTestModel.objects.annotate(
+            annotation=Case(
+                When(Exists(FKCaseTestModel.objects.filter(fk=OuterRef('pk'), integer=2)), then=Value("2")),
+                When(Exists(FKCaseTestModel.objects.filter(fk=OuterRef('pk'), integer=1)), then=Value("1")),
+                default=Value(""),
+                output_field=CharField(),
+            ),
+        ).get(integer=1)
+        self.assertEquals(instance.annotation, "1")
```
Comment (charettes) [inline tests/expressions_case/tests.py:324]: This is already covered in `expressions` tests

https://github.com/django/django/blob/9624703a06187060c1a494e533f3e27fed946de3/tests/expressions/tests.py#L749-L760

File: tests/expressions_case/tests.py [hunk 2]
```diff
@@ -312,6 +312,30 @@ def test_in_subquery(self):
...
+            annotation=Case(
+                When(Exists(FKCaseTestModel.objects.filter(fk=OuterRef('pk'), integer=1)), integer2=2,
+                     then=Value("2")),
+                When(Exists(FKCaseTestModel.objects.filter(fk=OuterRef('pk'), integer=1)), integer2=1,
+                     then=Value("1")),
+                default=Value(""),
+                output_field=CharField(),
+            ),
+        ).get(integer=1)
+        self.assertEquals(instance.annotation, "1")
```
Comment (charettes) [inline tests/expressions_case/tests.py:337]: `assertEquals` is deprecated in Python 3. Also not sure if this test belongs in `expressions.tests`.

Comment (rheard) [reply tests/expressions_case/tests.py:337]: Ok.

You're right, in light of the Exists test being in `expressions.tests`, this should probably go there too in its current form.

But I'm also not really trying to test `Exists` here, that just my use case. The real test should be using a query argument with a keyword argument, so I will modify to be simpler.

Comment (charettes) [reply tests/expressions_case/tests.py:337]: You could pass a literal boolean using `Value(True, output_field=models.BooleanField())`.

**Comment by felixxm** [comment]
@rheard Thanks for this patch :+1: I pushed minor edits, squashed commit, and added test for `TypeError`. Welcome aboard :boat: 

Please add a `versionchanged` annotation in [`When()` docs](https://docs.djangoproject.com/en/dev/ref/models/conditional-expressions/#when) and release notes in `3.2.txt`.

**Comment by rheard** [comment]
> @rheard Thanks for this patch 👍 I pushed minor edits, squashed commit, and added test for `TypeError`. Welcome aboard ⛵
> 
> Please add a `versionchanged` annotation in [`When()` docs](https://docs.djangoproject.com/en/dev/ref/models/conditional-expressions/#when) and release notes in `3.2.txt`.

@felixxm Thank you, I was not sure if i needed to squash commits or if it was done automatically. I'm not sure how to do that anyhow.

I have added that information as I **think** it should be formatted.

**Pull request merged**